### DISCUSSION TITLE
Increments in registry

### DISCUSSION THREAD

**Discussion by pavellety** [cirruslabs/tart#341] [Q&A] [answered]
Hi,

Let's say I build base image with packer and push it to the registry. This image contains basic stuff which worker needs. Then someone take this image and want to extend it with wget for example. So what's typically happened is that someone setup packer with vm_base_name equals to the original base image and build/push the image. Now we have two very similar images in the registry ~17GB each.

If we are talking about OCI compatible images these can be stored in layers. When I push two images when only difference is one additional binary like (wget). I would expect only increment is pushed to the registry not the whole image (17G).

Would it be technically possible to store only increments in registry?

Thanks

Regards,
Pavel

**Comment by edigaryev** [comment, marked as answer]
Unfortunately, it is not possible at the moment because `Virtualization.Framework` doesn't provide such functionality.

Currently, as a user of `Virtualization.Framework` we pass [an `URL` to a disk](https://developer.apple.com/documentation/virtualization/vzdiskimagestoragedeviceattachment) and that's it.

There's no way to implement your own `VZStorageDeviceAttachment` at the moment nor hook disk reads/writes (to store and apply diffs) in a way that is not considered a hack.

We could create a shadow VM disk though, and compare it when pushing to the original disk, but that would require 2x space and a way to figure out the VM's ancestor image.
